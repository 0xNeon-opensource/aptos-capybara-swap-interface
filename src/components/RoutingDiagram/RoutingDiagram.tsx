import { Currency, Percent } from '@uniswap/sdk-core'
import { FeeAmount } from '@uniswap/v3-sdk'
import Badge from 'components/Badge'
import CurrencyLogo from 'components/CurrencyLogo'
import DoubleCurrencyLogo from 'components/DoubleLogo'
import Row, { AutoRow } from 'components/Row'
import { Version } from 'hooks/useToggledVersion'
import { useTokenInfoFromActiveList } from 'hooks/useTokenInfoFromActiveList'
import { Box } from 'rebass'
import styled from 'styled-components/macro'
import { TYPE } from 'theme'
import { Z_INDEX } from 'theme'

import { ReactComponent as DotLine } from '../../assets/svg/dot_line.svg'

export interface RoutingDiagramEntry {
  percent: Percent
  path: [Currency, Currency, FeeAmount][]
  protocol: Version
}

const Wrapper = styled(Box)`
  align-items: center;
  background-color: ${({ theme }) => theme.bg0};
  width: 430px;
`

const RouteContainerRow = styled(Row)`
  display: grid;
  grid-gap: 8px;
  grid-template-columns: 24px 1fr 24px;
`

const RouteRow = styled(Row)`
  align-items: center;
  display: flex;
  justify-content: center;
  padding: 0.1rem 0.5rem;
  position: relative;
`

const PoolBadge = styled(Badge)`
  display: flex;
  padding: 0.25rem 0.5rem;
`

const DottedLine = styled.div`
  display: flex;
  align-items: center;
  position: absolute;
  width: calc(100%);
  z-index: 1;
  opacity: 0.5;
`

const DotColor = styled(DotLine)`
  path {
    stroke: ${({ theme }) => theme.bg4};
  }
`

const OpaqueBadge = styled(Badge)`
  background-color: ${({ theme }) => theme.bg2};
  display: grid;
  grid-gap: 4px;
  grid-auto-flow: column;
  justify-content: start;
  z-index: ${Z_INDEX.sticky};
`

const ProtocolBadge = styled(Badge)`
  background-color: ${({ theme }) => theme.bg3};
  border-radius: 0.3rem;
  z-index: ${Z_INDEX.sticky + 1};
`

const BadgeText = styled(TYPE.small)`
  word-break: normal;
`

export default function RoutingDiagram({
  currencyIn,
  currencyOut,
  routes,
}: {
  currencyIn: Currency
  currencyOut: Currency
  routes: RoutingDiagramEntry[]
}) {
  const tokenIn = useTokenInfoFromActiveList(currencyIn)
  const tokenOut = useTokenInfoFromActiveList(currencyOut)

  return (
    <Wrapper>
      {routes.map((entry, index) => (
        <RouteContainerRow key={index}>
          <CurrencyLogo currency={tokenIn} />
          <Route entry={entry} />
          <CurrencyLogo currency={tokenOut} />
        </RouteContainerRow>
      ))}
    </Wrapper>
  )
}

function Route({ entry: { percent, path, protocol } }: { entry: RoutingDiagramEntry }) {
  return (
    <RouteRow>
      <DottedLine>
        <DotColor />
      </DottedLine>
      <OpaqueBadge>
        <ProtocolBadge>
          <BadgeText fontSize={11}>{protocol.toUpperCase()}</BadgeText>
        </ProtocolBadge>
        <BadgeText fontSize={12} style={{ minWidth: 'auto' }}>
          {percent.toSignificant(2)}%
        </BadgeText>
      </OpaqueBadge>

      <AutoRow gap="1px" width="100%" style={{ justifyContent: 'space-evenly', zIndex: 2 }}>
        {path.map(([currency0, currency1, feeAmount], index) => (
          <Pool key={index} currency0={currency0} currency1={currency1} feeAmount={feeAmount} />
        ))}
      </AutoRow>
    </RouteRow>
  )
}

function Pool({ currency0, currency1, feeAmount }: { currency0: Currency; currency1: Currency; feeAmount: FeeAmount }) {
  const tokenInfo0 = useTokenInfoFromActiveList(currency0)
  const tokenInfo1 = useTokenInfoFromActiveList(currency1)

  return (
    <PoolBadge>
      <Box margin="0 5px 0 10px">
        <DoubleCurrencyLogo currency0={tokenInfo1} currency1={tokenInfo0} size={20} />
      </Box>
      <TYPE.small fontSize={12}>{feeAmount / 10000}%</TYPE.small>
    </PoolBadge>
  )
}
