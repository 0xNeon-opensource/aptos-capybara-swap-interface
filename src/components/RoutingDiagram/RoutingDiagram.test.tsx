import React from 'react'
import RoutingDiagram, { Route } from './RoutingDiagram'
import { Token } from '@uniswap/sdk-core'
import { FeeAmount } from '@uniswap/v3-sdk'
import { render } from 'test-utils'
import useHttpLocations from 'hooks/useHttpLocations'

const USDC = new Token(1, '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48', 6, 'USDC')
const DAI = new Token(1, '0x6B175474E89094C44Da98b954EedeAC495271d0F', 6, 'DAI')
const MKR = new Token(1, '0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2', 6, 'MKR')

const singleRoute: Route = { percent: 100, path: [[USDC, DAI, FeeAmount.LOW]] }

const multiRoute: Route[] = [
  { percent: 75, path: [[USDC, DAI, FeeAmount.LOW]] },
  {
    percent: 25,
    path: [
      [USDC, MKR, FeeAmount.MEDIUM],
      [MKR, DAI, FeeAmount.HIGH],
    ],
  },
]

// mock useHttpLocations to avoid having to load Web3
jest.mock('hooks/useHttpLocations')
const mockUseHttpLocation = useHttpLocations as jest.MockedFunction<typeof useHttpLocations>

// avoid large snapshots
jest.mock('components/DoubleLogo', () => 'DoubleLogo')

beforeEach(() => {
  mockUseHttpLocation.mockReturnValue([])
})

it('renders when no routes are provided', () => {
  const { asFragment } = render(<RoutingDiagram currencyIn={DAI} currencyOut={USDC} routes={[]} />)
  expect(asFragment()).toMatchSnapshot()
})

it('renders single route', () => {
  const { asFragment } = render(<RoutingDiagram currencyIn={USDC} currencyOut={DAI} routes={[singleRoute]} />)
  expect(asFragment()).toMatchSnapshot()
})

it('renders multi route', () => {
  const { asFragment } = render(<RoutingDiagram currencyIn={USDC} currencyOut={DAI} routes={multiRoute} />)
  expect(asFragment()).toMatchSnapshot()
})
