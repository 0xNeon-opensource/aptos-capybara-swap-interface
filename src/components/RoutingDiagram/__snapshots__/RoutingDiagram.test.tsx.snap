// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders multi route 1`] = `
<DocumentFragment>
  <div
    class="css-vurnku"
  >
    <div
      class="Column__AutoColumn-sc-1kykgp9-2 bXKOWF"
    >
      <div
        class="sc-bdnxRM Row-sc-nrd8cx-0 Row__AutoRow-sc-nrd8cx-3 fmeAVV drkpWa fRMqhK"
        width="auto"
      >
        <div
          class="Badge-sc-1mhw5si-0 gayll"
        >
          75
        </div>
        <div
          class="theme__TextWrapper-sc-18nh1jk-0 isUuLU css-1kt4f20"
        >
          USDC
        </div>
        <svg
          class="RoutingDiagram__StyledChevronRight-sc-o1ook0-0"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="9 18 15 12 9 6"
          />
        </svg>
        <div
          class="sc-bdnxRM Row-sc-nrd8cx-0 Row__AutoRow-sc-nrd8cx-3 fmeAVV drkpWa fRMqhK"
          width="auto"
        >
          <div
            class="Badge-sc-1mhw5si-0 gayll"
          >
            <div
              class="sc-bdnxRM Row-sc-nrd8cx-0 Row__AutoRow-sc-nrd8cx-3 fmeAVV drkpWa fRMqhK"
              width="auto"
            >
              <div
                class="DoubleLogo__Wrapper-sc-skatqk-0 diDXho"
              >
                <img
                  alt="USDC logo"
                  class="CurrencyLogo__StyledLogo-sc-1fvnadz-1 bzseWM DoubleLogo__HigherLogo-sc-skatqk-1 djPAFc"
                  src="https://raw.githubusercontent.com/uniswap/assets/master/blockchains/ethereum/assets/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48/logo.png"
                />
                <img
                  alt="DAI logo"
                  class="CurrencyLogo__StyledLogo-sc-1fvnadz-1 bzseWM DoubleLogo__CoveredLogo-sc-skatqk-2 jqDGQg"
                  sizeraw="16"
                  src="https://raw.githubusercontent.com/uniswap/assets/master/blockchains/ethereum/assets/0x6B175474E89094C44Da98b954EedeAC495271d0F/logo.png"
                />
              </div>
              500
            </div>
          </div>
          <svg
            class="RoutingDiagram__StyledChevronRight-sc-o1ook0-0"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="9 18 15 12 9 6"
            />
          </svg>
        </div>
        <div
          class="theme__TextWrapper-sc-18nh1jk-0 isUuLU css-1kt4f20"
        >
          DAI
        </div>
      </div>
    </div>
    <div
      class="Column__AutoColumn-sc-1kykgp9-2 bXKOWF"
    >
      <div
        class="sc-bdnxRM Row-sc-nrd8cx-0 Row__AutoRow-sc-nrd8cx-3 fmeAVV drkpWa fRMqhK"
        width="auto"
      >
        <div
          class="Badge-sc-1mhw5si-0 gayll"
        >
          25
        </div>
        <div
          class="theme__TextWrapper-sc-18nh1jk-0 isUuLU css-1kt4f20"
        >
          USDC
        </div>
        <svg
          class="RoutingDiagram__StyledChevronRight-sc-o1ook0-0"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="9 18 15 12 9 6"
          />
        </svg>
        <div
          class="sc-bdnxRM Row-sc-nrd8cx-0 Row__AutoRow-sc-nrd8cx-3 fmeAVV drkpWa fRMqhK"
          width="auto"
        >
          <div
            class="Badge-sc-1mhw5si-0 gayll"
          >
            <div
              class="sc-bdnxRM Row-sc-nrd8cx-0 Row__AutoRow-sc-nrd8cx-3 fmeAVV drkpWa fRMqhK"
              width="auto"
            >
              <div
                class="DoubleLogo__Wrapper-sc-skatqk-0 diDXho"
              >
                <img
                  alt="USDC logo"
                  class="CurrencyLogo__StyledLogo-sc-1fvnadz-1 bzseWM DoubleLogo__HigherLogo-sc-skatqk-1 djPAFc"
                  src="https://raw.githubusercontent.com/uniswap/assets/master/blockchains/ethereum/assets/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48/logo.png"
                />
                <img
                  alt="MKR logo"
                  class="CurrencyLogo__StyledLogo-sc-1fvnadz-1 bzseWM DoubleLogo__CoveredLogo-sc-skatqk-2 jqDGQg"
                  sizeraw="16"
                  src="https://raw.githubusercontent.com/uniswap/assets/master/blockchains/ethereum/assets/0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2/logo.png"
                />
              </div>
              3000
            </div>
          </div>
          <svg
            class="RoutingDiagram__StyledChevronRight-sc-o1ook0-0"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="9 18 15 12 9 6"
            />
          </svg>
        </div>
        <div
          class="sc-bdnxRM Row-sc-nrd8cx-0 Row__AutoRow-sc-nrd8cx-3 fmeAVV drkpWa fRMqhK"
          width="auto"
        >
          <div
            class="Badge-sc-1mhw5si-0 gayll"
          >
            <div
              class="sc-bdnxRM Row-sc-nrd8cx-0 Row__AutoRow-sc-nrd8cx-3 fmeAVV drkpWa fRMqhK"
              width="auto"
            >
              <div
                class="DoubleLogo__Wrapper-sc-skatqk-0 diDXho"
              >
                <img
                  alt="MKR logo"
                  class="CurrencyLogo__StyledLogo-sc-1fvnadz-1 bzseWM DoubleLogo__HigherLogo-sc-skatqk-1 djPAFc"
                  src="https://raw.githubusercontent.com/uniswap/assets/master/blockchains/ethereum/assets/0x9f8F72aA9304c8B593d555F12eF6589cC3A579A2/logo.png"
                />
                <img
                  alt="DAI logo"
                  class="CurrencyLogo__StyledLogo-sc-1fvnadz-1 bzseWM DoubleLogo__CoveredLogo-sc-skatqk-2 jqDGQg"
                  sizeraw="16"
                  src="https://raw.githubusercontent.com/uniswap/assets/master/blockchains/ethereum/assets/0x6B175474E89094C44Da98b954EedeAC495271d0F/logo.png"
                />
              </div>
              10000
            </div>
          </div>
          <svg
            class="RoutingDiagram__StyledChevronRight-sc-o1ook0-0"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="9 18 15 12 9 6"
            />
          </svg>
        </div>
        <div
          class="theme__TextWrapper-sc-18nh1jk-0 isUuLU css-1kt4f20"
        >
          DAI
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders single route 1`] = `
<DocumentFragment>
  <div
    class="css-vurnku"
  >
    <div
      class="Column__AutoColumn-sc-1kykgp9-2 bXKOWF"
    >
      <div
        class="sc-bdnxRM Row-sc-nrd8cx-0 Row__AutoRow-sc-nrd8cx-3 fmeAVV drkpWa fRMqhK"
        width="auto"
      >
        <div
          class="Badge-sc-1mhw5si-0 gayll"
        >
          100
        </div>
        <div
          class="theme__TextWrapper-sc-18nh1jk-0 isUuLU css-1kt4f20"
        >
          USDC
        </div>
        <svg
          class="RoutingDiagram__StyledChevronRight-sc-o1ook0-0"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="9 18 15 12 9 6"
          />
        </svg>
        <div
          class="sc-bdnxRM Row-sc-nrd8cx-0 Row__AutoRow-sc-nrd8cx-3 fmeAVV drkpWa fRMqhK"
          width="auto"
        >
          <div
            class="Badge-sc-1mhw5si-0 gayll"
          >
            <div
              class="sc-bdnxRM Row-sc-nrd8cx-0 Row__AutoRow-sc-nrd8cx-3 fmeAVV drkpWa fRMqhK"
              width="auto"
            >
              <div
                class="DoubleLogo__Wrapper-sc-skatqk-0 diDXho"
              >
                <img
                  alt="USDC logo"
                  class="CurrencyLogo__StyledLogo-sc-1fvnadz-1 bzseWM DoubleLogo__HigherLogo-sc-skatqk-1 djPAFc"
                  src="https://raw.githubusercontent.com/uniswap/assets/master/blockchains/ethereum/assets/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48/logo.png"
                />
                <img
                  alt="DAI logo"
                  class="CurrencyLogo__StyledLogo-sc-1fvnadz-1 bzseWM DoubleLogo__CoveredLogo-sc-skatqk-2 jqDGQg"
                  sizeraw="16"
                  src="https://raw.githubusercontent.com/uniswap/assets/master/blockchains/ethereum/assets/0x6B175474E89094C44Da98b954EedeAC495271d0F/logo.png"
                />
              </div>
              500
            </div>
          </div>
          <svg
            class="RoutingDiagram__StyledChevronRight-sc-o1ook0-0"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="9 18 15 12 9 6"
            />
          </svg>
        </div>
        <div
          class="theme__TextWrapper-sc-18nh1jk-0 isUuLU css-1kt4f20"
        >
          DAI
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders when no routes are provided 1`] = `
<DocumentFragment>
  <div
    class="css-vurnku"
  />
</DocumentFragment>
`;
